### DISCUSSION TITLE
vcf::AsyncWriter::into_inner() missing?

### DISCUSSION THREAD

**Discussion by holtgrewe** [zaeleus/noodles#210] [Q&A] [answered]
Hello, there is no `into_inner()` in `vcf::AsyncWriter`. Effectively, I cannot shutdown the inner writer which is necessary, for ensuring the output stream is flushed and properly shutdown, as far as I can see.

Am I missing something?

**Comment by holtgrewe** [comment]
Hm, unless I'm missing something, this even looks like a bug. When using a `#[tokio::main]` function and writing out many lines (say a few thousand) then the last batch will not be written out unless shutdown is called on the writer. It looks like the output file will be truncated in this case.

**Reply by zaeleus** [reply, marked as answer]
If you're using a buffered async writer, you have to call `flush`/`shutdown` before the writer gets dropped. See `tokio::io`'s "[Buffered Readers and Writers](https://docs.rs/tokio/latest/tokio/io/index.html#buffered-readers-and-writers)":

> However, you **must** flush [BufWriter](https://docs.rs/tokio/latest/tokio/io/struct.BufWriter.html) to ensure that any buffered data is written.
>
> ```rust
> // ...
>     } // Unless flushed or shut down, the contents of the buffer is discarded on drop.
> // ...
> ```

**Comment by zaeleus** [comment]
Underlying writer access for `vcf::AsyncWriter` was implemented in #211, and the methods are now available in noodles 0.55.0 / noodles-vcf 0.43.0.
